packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -65,6 +67,10 @@ export class ConnectionStateHandler {
                 this.applyForConnectedState("timeout");
             },
         );
+        // if we have a serialized clientId we will need to wait unless we see its "leave" op
+        if (serializedClientId !== undefined) {
```

Comment:
While connection state manager needs to receive new info (old clientId) in constructor, I believe adding new property makes it more complex and thus more error prone. Constructor should set up object as if we just disconnected (same way as do to right here).
But we should use existing properties only (including this.clientId) and move to (setup) some known existing state that already handles this condition, not create more states and reason about more complex state machine.
In other words, there should not be this.serializedClientId IMHO.

+ @markfields, who is poking at this code right now.